### ISSUE TITLE
Render Results differ between Windows and Linux of the same html file.

### ISSUE THREAD

**Issue by leiberumv** [wkhtmltopdf/wkhtmltopdf#2893]
I've getting a different result when rendering the html/php file in Windows compared to Linux. 

Using the exact same command/switches on windows wkhtmltopdf.exe version to linux wkhtmltopdf

Any ideas?

Screenshot result  compared here: 
![wkhtmltopdf_compare](https://cloud.githubusercontent.com/assets/9978891/14610704/e2cc30b6-0586-11e6-8928-2ef963057f6f.png)

**Comment by Getz85** [comment]
Same issue here, as if there was a zoom or a different resolution applied on Linux.

Did you solved your problem leiberumv?

**Comment by leiberumv** [comment]
@Getz85 No, although, to overcome this (not ideal) I'm having to output the PDF as a A0 size (over A4). this gives a good proportion and resolution, but the PDF is on a bigger print scale.

**Comment by Getz85** [comment]
Thanks, but I made it an other way: set the zoom parameters with a value less than 1 (0.78 in my case).

Not an ideal solution too, but at least I can keep the A4 format.

**Comment by lesliev** [comment]
This might be a font issue. See here: https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3023

**Comment by bunsen32** [comment]
Almost certainly to do with the Windows DPI setting… If you change the Windows desktop scaling (e.g., to 150%), log out, and log in again, and try again, your Windows PDF contents will shrink even further.

For my application I’ve explicitly set the ‘zoom’ setting to (windows DPI / 72)… (which seems like an unnecessary hassle).

There are a few other bugs on here which look like the same fundamental problem, e.g., #2991 

Ideally 10mm in CSS should be 10mm in the PDF, regardless of platform.

**Comment by pierregaste** [comment]
Same problem here...
I solved this by adding 2 properties:
- `--disable-smart-shrinking`
- `zoom 0.78` on my local machine (OSX) and `zoom 1` on my remote server (Debian)

**Comment by master455** [comment]
Same problem... 

I'm using Ubuntu 16.04 like server and Wkhtmltopdf 0.12.3 (because issues with 0.12.4 version). I developed report in Windows, but when install in server, report see not equal.

I apply the solution of @Getz85 (thanks!) add --zoom 0.78 and report is almost equal (95%).

**Comment by mohd-aidi** [comment]
@pierregaste  may i ask where should i put the properties?

- --disable-smart-shrinking
- zoom 0.78 on my local machine (OSX) and zoom 1 on my remote server - (Debian)

**Comment by mohd-aidi** [comment]
i think i figure it out. i use snappy and laravel so in my config/snappy.php i put it in options array
```
return array(
        'pdf' => array(
            'enabled' => true,
            'binary' => base_path(env('SNAPPY_WKHTMLTOPDF_PATH', 'vendor/wemersonjanuario/wkhtmltopdf-windows/bin/32bit/wkhtmltopdf.exe')),
            'timeout' => false,
            'options' => array(
                'zoom'  => 1
            ),
            'env'     => array(),
        ),
        'image' => array(
            'enabled' => true,
            'binary' => base_path(env('SNAPPY_WKHTMLTOIMAGE_PATH', 'vendor/wemersonjanuario/wkhtmltopdf-windows/bin/32bit/wkhtmltoimage.exe')),
            'timeout' => false,
            'options' => array(
                'zoom'  => 1
            ),
            'env'     => array(),
        ),
    );

```

**Comment by callaginn** [comment]
@mohd-aidi: Did you ever figure out a formula for calculating this on v0.12.4? I have three machines giving me drastically different results:
- Windows Server wo monitor (not controlled by me)
- 27in Retina iMac
- 27in non-Retina iMac

This is extremely frustrating, because the documentation doesn't explain much of anything. I keep finding random settings that work for one machine, but not another.

**Comment by mohd-aidi** [comment]
@callaginn I use php helper called PHP_OS to determine which setting will be used for each OS. Fortunately my case is between Windows and Ubuntu.

**Comment by pierregaste** [comment]
@mohd-aidi sorry for the reply time...
Actually, I use wkhtmltopdf with snappy too and I manually configure the zoom with environnement parameter. I set this options directly in the option's array, when I build the PDF.

**Comment by mohd-aidi** [comment]
@pierregaste i set it in config file so that i dont have to re-define it everytime i build the PDF.

**Comment by pierregaste** [comment]
@mohd-aidi Yes, of course 😃 
Setting variables in a config file (depending of your framework, project, ...) is a good practice.
But the point is that we have to configure an ENV variable depending of the OS.

**Comment by mohd-aidi** [comment]
@pierregaste i understand..
ok to clear things up, this is how i configure if the OS is windows or ubuntu..

this configuration is only between my 2 servers which is Ubuntu and Windows Server 2013

`<?php
if(PHP_OS == 'WINNT'){
    return array(
        'pdf' => array(
            'enabled' => true,
            'binary' => base_path(env('SNAPPY_WKHTMLTOPDF_PATH', 'vendor/wemersonjanuario/wkhtmltopdf-windows/bin/32bit/wkhtmltopdf.exe')),
            'timeout' => false,
            'options' => array(
                'zoom'  => 1,
                'page-size'    => 'A4'
            ),
            'env'     => array(),
        ),
        'image' => array(
            'enabled' => true,
            'binary' => base_path(env('SNAPPY_WKHTMLTOIMAGE_PATH', 'vendor/wemersonjanuario/wkhtmltopdf-windows/bin/32bit/wkhtmltoimage.exe')),
            'timeout' => false,
            'options' => array(
                'zoom'  => 1
            ),
            'env'     => array(),
        ),
    );

}else{
    return array(
        'pdf' => array(
            'enabled' => true,
            'binary' => base_path(env('SNAPPY_WKHTMLTOPDF_PATH', 'vendor/wemersonjanuario/wkhtmltopdf-windows/bin/32bit/wkhtmltopdf.exe')),
            'timeout' => false,
            'options' => array(
                'disable-smart-shrinking' => true,
                'zoom'  => 0.575
            ),
            'env'     => array(),
        ),
        'image' => array(
            'enabled' => true,
            'binary' => base_path(env('SNAPPY_WKHTMLTOIMAGE_PATH', 'vendor/wemersonjanuario/wkhtmltopdf-windows/bin/32bit/wkhtmltoimage.exe')),
            'timeout' => false,
            'options' => array(
                'disable-smart-shrinking' => true,
                'zoom'  => 0.575
            ),
            'env'     => array(),
        ),
    );
}
`

**Comment by glanes** [comment]
Just want to add my contribution as I had the same problem.
use @pierregaste suggestion
knp_snappy:
    pdf:
        enabled:    true
        binary:     %kernel.root_dir%/../vendor/h4cc/wkhtmltopdf-amd64/bin/wkhtmltopdf-amd64 --disable-smart-shrinking --zoom 1
        options:    []

**Comment by Tomsgu** [comment]
Does this problem still appears in 0.12.5?

**Comment by glanes** [comment]
> Does this problem still appears in 0.12.5?

The version that I am using is 0.12.4, but I have updated the wkhtmltopdf-amd64 for the latest version when I was testing.

**Comment by jeremielodi** [comment]
I had this issue too, so I decided to install  <a href="https://www.ostechnix.com/install-microsoft-windows-fonts-ubuntu-16-04/">windows fonts on ubuntu</a> 
The result was almost what I was expected

**Comment by Tomsgu** [comment]
Difference between mac os and linux should be fixed in 0.12.5 See #3241. And the DPI differences were fixed in 0.12.4 See: #2463.

As @jeremielodi mentioned the same result cannot be achieved if you use a different fonts.

Is this really fixed in a newest version? If not can you add an example in version 0.12.5?
